### PR TITLE
feat!: update Yoga to 3.0

### LINKED ISSUE #2507: [solution] __dirname is not defined in ES module scope
**Describe the bug**

When building react-pdf in ESM mode, you may encounter an error:

```
__dirname is not defined in ES module scope
```

This is because yoga-layout uses `__dirname` in their code. This has already been reported to them: https://github.com/facebook/yoga/issues/1559

To work around this issue, for now, consider using `inject` option in esbuild:

`cjs-shim.ts`
```ts
import { createRequire } from 'node:module';
import path from 'node:path';
import url from 'node:url';

globalThis.require = createRequire(import.meta.url);
globalThis.__filename = url.fileURLToPath(import.meta.url);
globalThis.__dirname = path.dirname(__filename);
```

`esbuild.ts`
```ts
await esbuild.build({
  // …
  inject: ['cjs-shim.ts'],
  // …
});
```

If you're not using esbuild, you may manually import the `cjs-shim.ts` file BEFORE react-pdf is loaded.

### LINKED ISSUE #2692: Update Yoga to 3.0
Updating Yoga to 3.0 will bring several features into React-PDF:

- Support for position: static
- Support for align-content: space-evenly
- Improvements to layout correctness
- Yoga’s JavaScript bindings are now distributed as an ES Module
- Fixes several crashes in Java bindings
- Some existing Yoga APIs have been removed

Source: https://www.yogalayout.dev/blog/announcing-yoga-3.0

But also, it will:

- Resolve https://github.com/diegomura/react-pdf/issues/2507, bringing in true, zero configuration ESM compatibility 

Potential breaking changes:

- Yoga 3.0 is distributed as ESM-only. We might be forced to make React-PDF ESM-only as well.
- Given that there are "layout correctness" improvements makes me believe this should be considered a breaking change, as people may have used certain workarounds that would stop working after an update.

### PR DESCRIPTION
**Pull request by wojtekmaj**
Closes #2692
Closes #2507

This PR updates Yoga to 3.0.

See the list of new features, improvements, and potential breaking changes in #2692.

### CODE REVIEW

**Comment by diegomura** [comment]
@wojtekmaj this looks good. What's missing?

**Comment by wojtekmaj** [comment]
@diegomura See the original ticket: we will be unable to provide CJS builds as Yoga 3 is a pure ESM module.

**Comment by diegomura** [comment]
Sorry! Missed that. Huh, good decision to make. Would be nice to push forward that and just embrace ESM but I'm afraid that would break a lot of setups. Will think about this. Any input would be welcomed!

**Comment by wojtekmaj** [comment]
I think it's really up to you: how much further are you able (and willing to!) push the v2 forward? If you think that significant improvements can still be made, you may want to hold off with v3 for a while, to avoid maintaining two releases at the same time. If you think v2 is "pretty much ready" and you're basically only _maintaining_ it, then it would be best to bite the bullet and remove CJS builds.

Please note that removing CJS builds does not stop CJS users from:
- Using v2 🤷 They clearly like outdated solutions anyway
- Importing ESM modules into CJS asynchronously (although that's a bit of a burden)
- Using Node with `--experimental-require-module` flag which _may_ just work
- Using a bundler like Rollup before running code in Node to benefit from their built-in ESM/CJS interop to avoid all module resolution issues

**Comment by karlhorky** [comment]
I wonder if Yoga v2 with its unusual CommonJS wrapper code is also causing issues like this property access error `Yoga.Overflow.Hidden` that we've been experiencing:

- https://github.com/diegomura/react-pdf/issues/2499#issuecomment-2166770560

**Comment by karlhorky** [comment]
@wojtekmaj is there an easy way to try this PR out?

If this isn't published somewhere already, would it be possible to get a prerelease version published?

I'm currently patching `@react-pdf/layout` in `node_modules` and using pnpm Overrides to force `yoga-layout@3.0.4`, and it appears to be working, but it's a bit brittle...

**Comment by wojtekmaj** [comment]
I think this is the best we can do until it's released :(

File: packages/layout/src/yoga/index.js [hunk 1]
```diff
@@ -1,14 +1,13 @@
 /* eslint-disable import/prefer-default-export */
 
-import * as Yoga from 'yoga-layout';
+import Yoga from 'yoga-layout';
 
 export const loadYoga = async () => {
```
Comment (michaelgmcd) [inline packages/layout/src/yoga/index.js:5]: Looked a little bit and it has some trickle-down changes needed, but this doesn't need to be `async` anymore, correct?

Comment (wojtekmaj) [reply packages/layout/src/yoga/index.js]: Good eye. Thanks!

**Comment by michaelgmcd** [comment]
It may actually be more beneficial to go back to keeping things async and using the `yoga-layout/load` export. The default one requires top-level await (may break some users):
https://github.com/facebook/yoga/tree/main/javascript#requirements

**Comment by wojtekmaj** [comment]
Thanks for the heads up. PDF.js introduced top-level await recently and they rolled it back after a huge backslash, so I guess you'd be right.

**Comment by seanroddy-fbg** [comment]
Hi, is there an expected release data for this feature?

**Comment by diegomura** [comment]
I'm very tempted on removing CJS support. We should look forward. I feel it's kind of our responsibility to help things moving forward as lib authors. Newer solutions won't become adopted if people do not have an incentive to upgrade. Also, given bandwidth of lib authors feels like it simplifies things. @wojtekmaj curious what you think

**Comment by wojtekmaj** [comment]
I can't agree more. And frankly, it's not even a question if we want to move react-pdf forward. More and more libs are abandoning CJS, like our core dependencies.

**Comment by diegomura** [comment]
Amazing. I'll remove CJS support and we can land this after then!

**Comment by diegomura** [comment]
Done in https://github.com/diegomura/react-pdf/pull/2871

**Review by diegomura — APPROVED** [review]
utACK

**Pull request merged**